src/mission_companion.cpp

```diff
@@ -258,10 +299,476 @@ bool talk_function::outpost_missions( npc &p, std::string id, std::string title
             }
             entry = entry + _("\n \nDo you wish to bring your allies back into your party?");
             col_missions["Recover Ally from Carpentry Work"] = entry;
-            keys.push_back("Recover Ally from Carpentry Work");
+            mission_key_push( key_vectors, "Recover Ally from Carpentry Work" );
+        }
+    }
+
+    //Used to determine what kind of OM the NPC is sitting in to determine the missions and upgrades
+    const point omt_pos = ms_to_omt_copy( g->m.getabs( p.posx(), p.posy() ) );
+    oter_id &omt_ref = overmap_buffer.ter( omt_pos.x, omt_pos.y, p.posz() );
+    std::string om_cur = omt_ref.id().c_str();
+    std::string bldg;
+    std::vector<std::pair<std::string, tripoint>> om_expansions = om_building_region( p, 1, true );
+
+    if( id == "FACTION_CAMP" ){
```

Comment:
You do a lot of `id == something` checks here.
You should extract the guts of those `if`s to functions. If you need some context (in this case `col_missions` etc.), you should either make those functions methods of some class, or pass them a `struct` that has all those guts.